lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+    New.CustomName() = CurrentArgument.CustomName();
+    New.OriginalName() = CurrentArgument.OriginalName();
+
+    CurrentAlignment = NextAlignment;
+  }
+
+  // And don't forget the very last field.
+  const model::StructField &LastArgument = *std::prev(Types.Fields().end());
+  model::Argument &New = Result.emplace_back();
+  New.Index() = IndexOffset++;
+  New.Type() = LastArgument.Type();
+  New.CustomName() = LastArgument.CustomName();
+  New.OriginalName() = LastArgument.OriginalName();
+
+  // Make sure the size does not contradict the final alignment.
+  uint64_t FullAlignment = *ABI.alignment(StackArgumentTypes);
+  if (!isZeroOrPowerOfTwo(FullAlignment)) {
+    revng_log(Log,
+              "The natural alignment of a type is not a power of two:\n"
+                << serializeToString(Types));
+    return std::nullopt;
+  }
+
+  revng_assert(Types.Size() % FullAlignment == 0);
+
+  return Result;
+}
+
+using ModelRegister = model::Register::Values;
+template<typename RegisterType>
+std::optional<model::QualifiedType>
+TDManager::tryConvertingReturnValue(const SortedVector<RegisterType> &Used) {
+  if (Used.size() == 0) {
+    // The function doesn't return anything.
+    return model::QualifiedType{
+      getPrimitiveType(model::PrimitiveTypeKind::Void, 0), {}
+    };
+  }
+
+  const auto Architecture = model::ABI::getArchitecture(ABI.ABI());
+
+  // The case where the only used register is the pointer-to-copy location
+  // needs to be handled in a special way.
+  if (Used.size() == 1
+      && Used.begin()->Location() == ABI.ReturnValueLocationRegister()) {
+    if (Used.begin()->Type().isPointer()) {
+      // If the type is already a pointer, preserve it.
+      return Used.begin()->Type();
+    } else if (!Used.begin()->Type().isPrimitive()) {
+      // If type is a primitive, rewrite it with an arbitrary pointer type.
+      auto Byte = getPrimitiveType(model::PrimitiveTypeKind::Generic, 8);
+      return model::QualifiedType(Byte, {}).getPointerTo(Architecture);
+    } else {
+      revng_log(Log,
+                "Abort the conversion since proceeding would mean discarding "
+                "an arbitrary non-scalar type.");
+      return std::nullopt;
+    }
+  }
+
+  // The return value is register-based, deal with it similarly to how the
+  // arguments are worked with, which requires relying on register state
+  // deduction and ordering.
+  abi::RegisterState::Map Map(model::ABI::getArchitecture(ABI.ABI()));
+  for (const RegisterType &Register : Used)
+    Map[Register.Location()].IsUsedForReturningValues = abi::RegisterState::Yes;
+  auto Enforced = abi::enforceRegisterStateDeductions(Map, ABI);
+
+  llvm::SmallSet<model::Register::Values, 8> ArgumentRegisters;
+  for (auto [Register, Pair] : Enforced)
+    if (abi::RegisterState::shouldEmit(Pair.IsUsedForPassingArguments))
+      ArgumentRegisters.insert(Register);
+  auto Ordered = abi::orderArguments(ArgumentRegisters, ABI);
```

Comment:
Ah, whoops, it's a bad copy-paste.

I probably didn't run into any explicit problems because many ABIs (for example ARMs) use the same registers for both returning values and accepting arguments.

Reply:
This is done, see the `(not reviewed)` commits.